woocommerce/woocommerce-gutenberg-products-block
# Fix translations, bundle complete JS translation files

Fixes #336 – This is a bit complicated and not amazingly well documented, so this PR was a little bit trial-and-error.

- We use the WordPress.org translation system to bundle translation files when a site is set to non US-English locale. It recently learned how to read JS files for translation, but before that you needed another PHP file with all the JS strings pulled out - this is what `languages/woo-gutenberg-products-block.php` is for. This PR updates this file with all the new strings, but is probably unnecessary.
- As of WP 5.0, we can load in translations for JS files using `wp_set_script_translations`. We need to do this for every JS file, not once per textdomain. This PR **adds this function to each script registration,** and additionally **moves the style registration closer to the script** reg, so that everything for a single block is in one place.
- Unfortunately, [it seems that wp.org isn't generating the JSON files needed for the translations.](https://meta.trac.wordpress.org/ticket/3876) Until that's the case, we can bundle the complete translations in the plugin. Only Danish and French are complete, so those are the translations added. We also need to add one JSON file per JS file, the translation name includes the MD5 hash of the JS filename. Ideally these would only contain the strings used in that file, but I don't have an automated method for that. **So we've added 8 copies of the same JSON,** one for each block JS file.
- The legacy block is doing translations wrong (no textdomain), so it will not appear translated.

An example of a French translated block:
![screen shot 2019-01-16 at 8 39 31 pm](https://user-images.githubusercontent.com/541093/51289614-da6c5180-19ce-11e9-8180-8cf8b1000ba5.png)

### How to test the changes in this Pull Request:

1. Set your site to Danish or French
2. Go to the editor
3. Check that there are no relevant console errors
4. Add each block– note that if you, like me, use the `/prod…` command to enter products, that won't work in Danish because the translated name is not "Products" 😆
5. Check that all strings are translated.
6. View a post with the legacy block, check that it's not broken (but it won't be translated).

For reference, a list of all blocks:
- Best Selling Products
- Featured Product
- Hand-picked Products
- Newest Products
- On Sale Products
- Products by Category
- Top Rated Products

Note: In a future PR we can remove the `i18n:php` command & the `languages/woo-gutenberg-products-block.php` file – but for now let's leave it as-is.
